# genomisc-training

## Miscellaneous materials (tutorials, exercises, testdata) to develop essential 'survival'-skills in bioinformatics.

git clone..

contact: c.hahn@hull.ac.uk

1. [Introduction](https://github.com/HullUni-bioinformatics/genomisc-training#introduction)
  1. Reproducible science
  2. The command line 
    1. Basics
    2. writing/running scripts
  3. Docker
  5. Using HPC
    1. Connecting
    2. Job submission
      1. SLURM
      2. PBS
2. [Manipulating FASTQ data](https://github.com/HullUni-bioinformatics/genomisc-training#manipulating-fastq-data)
  1. FASTQ basics
  2. FASTQ file manipulation using command line skills
  3. FASTQ trimming
    1. Fastx-toolkit
    2. Trimmomatic
  4. Error correction
    1. Illumina data
    2. PacBio data
  5. paired end read merging
    1. FLASh
    2. Pear
  6. Demultiplexing
3. [Read mapping](https://github.com/HullUni-bioinformatics/genomisc-training#read-mapping)
  1. BWA
  2. Bowtie
4. [RADseq data](https://github.com/HullUni-bioinformatics/genomisc-training#radseq-data)
  1. Stacks
  2. Pyrad
5. [SNP calling](https://github.com/HullUni-bioinformatics/genomisc-training#snp-calling)
  1. Freebayes
6. [SNP annotation](https://github.com/HullUni-bioinformatics/genomisc-training#snp-annotation)
  1. SNPeff
7. [Genome assembly](https://github.com/HullUni-bioinformatics/genomisc-training#genome-assembly)
  1. Illumina data
    1. Velvet
    2. Spades
    3. Celera
  2. PacBio
    1. CANU
    2. FALCON
    3. MIRA
  3. Hybrid
    1. MIRA
    2. Spades
    3. Celera
  4. Metaassembly, Gapfilling and polishing
    1. PBjelly
    2. quickmerge
  3. Assembly evaluation
    1. Basic stats
    2. Completeness
    3. Contamination
8. [RNAseq data](https://github.com/HullUni-bioinformatics/genomisc-training#rnaseq-data)
  1. Denovo
  2. Reference genome based
9. [metabarcoding](https://github.com/HullUni-bioinformatics/genomisc-training#metabarcoding)
  1. A basic BLAST search
  2. MEGAN
10. Structural genome annotation
11. Functional genome annotation

## Introduction

## Manipulating FASTQ data

# FASTQ file manipulation using command line skills

The UNIX command line provides highly efficient, simple and incredibly powerful tools for text file manipulation. Much of the NGS data you will be processing are nothing more than text files. Proficiency with some of these basic tools will get you a long way (you may not need anything else), so I have prepared a number of simple exercises that should help you develop your command line skills, specifically in the context of FASTQ data manipulation. It will hopefully also help you to get a feel for the kind of data that you will be working with. Get started [here]().


## Read mapping

## RADseq data

## SNP calling

## SNP annotation

## Genome assembly

## RNAseq data

## metabarcoding
